### ISSUE TITLE
no-useless-escape false positives on JSX strings

### ISSUE THREAD

**Issue by turadg** [eslint/eslint#5882]
This is a JSX issue, but since it's in `eslint/no-useless-escape` I thought it would be appropriate to file here.

**What version of ESLint are you using?**
2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**
Babel-ESLint

**Please show your full configuration:**
It's super long but these should be enough,
```js
module.exports = {

  // base off Airbnb Javascript Style Guide
  // https://github.com/airbnb/javascript
  'extends': [
    'eslint-config-airbnb',
  ],

  'parser': 'babel-eslint',
}
```

**What did you do? Please include the actual source code causing the issue.**
```jsx
<input
  autoComplete="cc-cvc cc-csc"
  maxLength="4"
  pattern="[\d]{3,4}"
  placeholder="CVC"
/>
```

**What did you expect to happen?**

The pattern `"[\d]{3,4}"` would not be flagged as the backslash is not an escape character in this context. (It's a literal backslash, like in HTML.)

**What actually happened? Please include the actual, raw output from ESLint.**

```
  135:21  error  Unnecessary escape character: \d  no-useless-escape
```

**Comment by platinumazure** [comment]
I don't know JSX-- do their "attribute strings" have the same escape syntax and semantics as strings?

If so, then `\d` is actually incorrect and you should be using `\\d`, and the rule is flagging correctly. But if JSX has different syntax and semantics for string escapes, then yes, there might be a bug or enhancement request here.

**Comment by turadg** [comment]
JSX treats attribute strings the way HTML does.

https://jsfiddle.net/69z2wepo/39157/

**Comment by platinumazure** [comment]
Fascinating. espree does return the string as a literal with the backslash properly escaped, according to ASTExplorer: http://astexplorer.net/#/92TrD2yHtJ

Can you `npm ls` so we can see what version of espree you're running with?

**Comment by turadg** [comment]
Very handy! I changed the parser from "acorn" to "babel-eslint" and the value literal is still `"[\\d]{3,4}"`. (Fwiw, same for Espree, but Typescript parses as `"[d]{3,4}"`.)

I also commented out `babel-eslint` in my eslintrc and it made no difference.

**Comment by platinumazure** [comment]
Ah well, worth a shot. Thanks for trying those out. I'll dig in further
when I get home.

**Comment by mysticatea** [comment]
I saw the spec: https://facebook.github.io/jsx/
JSXText seems to not have any escape sequence.

And I confirmed this with espree, too.
I will fix this.

```
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> cat test.js
<foo attr="\d"/>;
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> eslint -v
v2.8.0
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> eslint test.js --rule no-useless-escape:2 --parser-options "ecmaFeatures:{jsx:true}" --no-ignore --no-eslintrc

C:\Users\t-nagashima.AD\Documents\GitHub\eslint\test.js
  2:12  error  Unnecessary escape character: \d  no-useless-escape

✖ 1 problem (1 error, 0 warnings)
